src/sendqueue.rs

```diff
@@ -0,0 +1,93 @@
+use std::convert::TryInto;
+use std::ptr::NonNull;
+use std::time::{SystemTime, UNIX_EPOCH};
+
+use libc::{c_int, c_uint};
+
+use crate::raw;
+use crate::Error;
+use crate::{Active, Capture};
+
+pub struct SendQueue {
+    squeue: NonNull<raw::pcap_send_queue>,
+    sync: c_int,
+}
+
+impl SendQueue {
+    pub fn new(memsize: c_uint) -> Result<Self, Error> {
+        let squeue = unsafe { raw::pcap_sendqueue_alloc(memsize) };
+        let squeue = if let Some(squeue) = NonNull::new(squeue) {
+            squeue
+        } else {
+            return Err(Error::InsufficientMemory);
+        };
+
+        Ok(Self { squeue, sync: 0 })
+    }
+
+    pub fn maxlen(&self) -> c_int {
+        unsafe { (*self.squeue.as_ptr()).maxlen }
+    }
+
+    pub fn len(&self) -> c_int {
+        unsafe { (*self.squeue.as_ptr()).len }
+    }
+
+    pub fn sync(&mut self, sync: bool) {
+        self.sync = if sync { 1 } else { 0 }
+    }
+
+    /// Add a packet to the queue.
+    pub fn queue(&mut self, buf: &[u8]) -> Result<(), Error> {
+        let start = SystemTime::now();
```

Comment:
do we want the user have more control over what we use as time ?

Reply:
The reason `sync` exists is basically to be able to send out packets with a high precision delay.  One might imagine that the use-case is primarily to perform replays/simulations.  One could further argue _if_ one would use this feature, it is likely that one would be loading packets from a capture, and would want to recreate the timings in the sendqueue from that capture buffer.  From that perspective you are completely right.

My use-case is not that though -- I'm interested in sending packets _**very**_ fast, and thus I do not use `sync` at all.  And I imagine that the replay case is far more niche than "send packets very fast".

Here's a suggestion:
- Add an `ts: Option<(i32, i32)>` for a timestamp to `queue()`
- if `sync` is set and:
  - `ts` is None, then get the system time is fetched as it is currently.
  - `ts` is Some, use the tuple's values as sec and usec instead.
- if `sync` is not set, then don't get time at all -- populate the time with zeroes.

The prior art:ish thing to do is to take in a pkthdr buffer to `queue`, since (as I noted above) that _if_ one uses this feature it is likely that one is populating the sendqueue from a capture buffer.  However, I _really_ want to set the lengths from the input slice, and thus the only interesting part of the header is the timestamp.

Reply:
I think this time thing is a problem, first the code will stop working in something like ~15.5 years, secondly the user could want use monotonic time or whatever. https://github.com/rust-lang/libc/issues/1848 at the very least you should cast it to `time_t` and `suseconds_t` and we should probably let the user give us a `Duration`

Reply:
The seconds in `timeval` is an `i32` in `libc` on Windows for some reason.  The wraparound is unlikely to ever be an issue if purely relative times are used.  Switched to use `Duration` as you suggested, which makes specifying relative times much more natural.